build-llvm.py

```diff
@@ -0,0 +1,438 @@
+                        powerpc, and x86 backends because that's what the Linux kernel is currently concerned with.
+                        If you would like to override this, you can use this parameter and supply a list that is
+                        supported by LLVM_TARGETS_TO_BUILD: https://llvm.org/docs/CMake.html#llvm-specific-variables
+
+                        Example: -t "AArch64;X86"
+
+                        """),
+                        type=str,
+                        default="AArch64;ARM;PowerPC;X86")
+    return parser.parse_args()
+
+
+def linker_test(cc, ld):
+    """
+    Test to see if the supplied ld value will work with cc -fuse=ld
+    :param cc: A working C compiler to compile the test program
+    :param ld: A linker to test -fuse=ld against
+    :return: 0 if the linker supports -fuse=ld, 1 otherwise
+    """
+    echo = subprocess.Popen(['echo', 'int main() { return 0; }'],
+                            stdout=subprocess.PIPE)
+    cc_call = subprocess.Popen(
+        [cc, '-fuse-ld=' + ld, '-o', '/dev/null', '-x', 'c', '-'],
+        stdin=echo.stdout,
+        stderr=subprocess.DEVNULL)
+    cc_call.communicate()
+    return cc_call.returncode
+
+
+def check_cc_ld_variables():
+    """
+    Sets the cc, cxx, and ld variables, which will be passed to cmake
+    :return: A tuple of valid cc, cxx, ld values that can be used to compile LLVM
+    """
+    utils.print_header("Checking CC and LD")
+    cc, linker, ld = None, None, None
+    # If the user didn't supply a C compiler, try to find one
+    if 'CC' not in os.environ:
+        possible_compilers = ['clang-9', 'clang-8', 'clang-7', 'clang', 'gcc']
+        for compiler in possible_compilers:
+            cc = shutil.which(compiler)
+            if cc is not None:
+                break
+        if cc is None:
+            raise RuntimeError(
+                "Neither gcc nor clang could be found on your system!")
+    # Otherwise, get its full path
+    else:
+        cc = shutil.which(os.environ['CC'])
+
+    # Evaluate if CC is a symlink. Certain packages of clang (like from
+    # apt.llvm.org) symlink the clang++ binary to clang++-<version> in
+    # /usr/bin, which then points to something like /usr/lib/llvm-<version/bin.
+    # This won't be found by the dumb logic below and trying to parse and figure
+    # out a heuristic for that is a lot more effort than just going into the
+    # folder that clang is actually installed in and getting clang++ from there.
+    cc = os.path.realpath(cc)
+    cc_folder = os.path.dirname(cc)
+
+    # If the user didn't supply a C++ compiler
+    if 'CXX' not in os.environ:
+        if "clang" in cc:
+            cxx = "clang++"
+        else:
+            cxx = "g++"
+        # Use the one that is located where CC is
+        cxx = shutil.which(cxx, path=cc_folder + ":" + os.environ['PATH'])
+    # Otherwise, get its full path
+    else:
+        cxx = shutil.which(os.environ['CXX'])
+    cxx = cxx.rstrip()
```

Comment:
If whitespace is not important in either side, I would simply do `strip()` instead. Otherwise it makes me wonder if the left-side whitespace is important, but maybe that is just me :-)